SYCL/USM/copy.cpp

```diff
@@ -12,83 +12,30 @@
 // RUN: %GPU_RUN_PLACEHOLDER %t1.out
 // RUN: %ACC_RUN_PLACEHOLDER %t1.out
 
-#include <sycl/sycl.hpp>
+#include "copy.hpp";
 
 using namespace sycl;
 using namespace sycl::usm;
 
-template <typename T> class transfer;
-
-static constexpr int N = 100; // should be even
-
 struct test_struct {
   short a;
   int b;
   long c;
   long long d;
   half e;
   float f;
-  double g;
 };
 
 bool operator==(const test_struct &lhs, const test_struct &rhs) {
   return lhs.a == rhs.a && lhs.b == rhs.b && lhs.c == rhs.c && lhs.d == rhs.d &&
-         lhs.e == rhs.e && lhs.f == rhs.f && lhs.g == rhs.g;
-}
-
-template <typename T> T *regular(queue q, alloc kind) {
-  return malloc<T>(N, q, kind);
-}
-
-template <typename T> T *aligned(queue q, alloc kind) {
-  return aligned_alloc<T>(alignof(long long), N, q, kind);
-}
-
-template <typename T> void test(queue q, T val, T *src, T *dst, bool dev_dst) {
-  q.fill(src, val, N).wait();
-
-  // Use queue::copy for the first half and handler::copy for the second
-  q.copy(src, dst, N / 2).wait();
-  q.submit([&](handler &h) { h.copy(src + N / 2, dst + N / 2, N / 2); }).wait();
-
-  T *out = dst;
-
-  std::array<T, N> arr;
-  if (dev_dst) { // if copied to device, transfer data back to host
-    buffer buf{arr};
-    q.submit([&](handler &h) {
-      accessor acc{buf, h};
-      h.parallel_for<transfer<T>>(N, [=](id<1> i) { acc[i] = dst[i]; });
-    });
-    out = arr.data();
-  }
-
-  for (int i = 0; i < N; ++i) {
-    assert(out[i] == val);
-  }
-
-  free(src, q);
-  free(dst, q);
-}
-
-template <typename T> void runTests(queue q, T val, alloc kind1, alloc kind2) {
-  bool dev_dst1 = (kind1 == alloc::device);
```

Comment:
Why isn't it enough to make early returns from this function, without creating an extra .cpp file? Same for other places.

Reply:
Not sure you are asking for `dev_dst1 = (kind1 == alloc::device);`?
I guess you mean returns from this function if the device `!has(sycl::aspect::fp64)`?
The JIT compiler raises unsupported 'double' type error in compiling no matter we have `if/else` statement to return early. We have to split them into two TUs.

Reply:
How do customers solve this? Imagine I have fast double algorithm and slow float algorithm, and I want to do host-based dispatch to run one based on the current device. Are you saying this is impossible?

Reply:
Do we have compiler implementation lagging behind? Probably not everything from SYCL2020 in this regard is implemented yet.

+ @gmlueck

Reply:
Yes, this is a bug in DPC++, and we have had customer complaints about it.  Applications should not be required to put their "double" kernels in a separate source file.  The SYCL 2020 specification requires that this work even when there is a single source file.  The proposed solution is the design that is outlined in [OptionalDeviceFeatures](https://github.com/intel/llvm/blob/sycl/sycl/doc/design/OptionalDeviceFeatures.md).

Reply:
@AlexeySachkov do we plan to address it soon? Otherwise we should split to keep testing specific underlying features properly.

Reply:
I still think splitting like this increases maintainability burden greatly. I'd rather see something like:

```
// test.cpp
// RUN: ... -DENABLE_FP64=false
constexpr bool EnableFP64 = ENABLE_FP64;

void test() {
   // non-fp64-case
   do_smth();

   if constexpr (EnableFP64)
     do_smth64();
}

// test-fp64.cpp
; RUN: ... -DENABLE_FP64=true
#include <test-fp.cpp>
```

Reply:
@aelovikov-intel, that's good solution!
The preprocess step is able to strip out those fp64 code, and no complain is raised from GPU compiler `error: Double type is not supported on this platform.` any more.
I will update this change in 24 files `<test>_aspect-fp64.cpp` very soon.

Reply:
Thanks!